src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostOnAutoInsertEndpoint.cs

```diff
@@ -0,0 +1,165 @@
+                Method = VSInternalMethods.OnAutoInsertName,
+                RegisterOptions = new VSInternalDocumentOnAutoInsertOptions()
+                    .EnableOnAutoInsert(triggerCharacters)
+            };
+        }
+
+        return null;
+    }
+
+    protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(VSInternalDocumentOnAutoInsertParams request)
+        => request.TextDocument.ToRazorTextDocumentIdentifier();
+
+    protected override async Task<VSInternalDocumentOnAutoInsertResponseItem?> HandleRequestAsync(VSInternalDocumentOnAutoInsertParams request, RazorCohostRequestContext context, CancellationToken cancellationToken)
+    {
+        var razorDocument = context.TextDocument.AssumeNotNull();
+
+        _logger.LogDebug($"Resolving auto-insertion for {razorDocument.FilePath}");
+
+        var clientSettings = _clientSettingsManager.GetClientSettings();
+        var enableAutoClosingTags = clientSettings.AdvancedSettings.AutoClosingTags;
+        var formatOnType = clientSettings.AdvancedSettings.FormatOnType;
+        var indentWithTabs = clientSettings.ClientSpaceSettings.IndentWithTabs;
+        var indentSize = clientSettings.ClientSpaceSettings.IndentSize;
+
+        _logger.LogDebug($"Calling OOP to resolve insertion at {request.Position} invoked by typing '{request.Character}'");
+        var data = await _remoteServiceInvoker.TryInvokeAsync<IRemoteAutoInsertService, Response>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken)
+                => service.TryResolveInsertionAsync(
+                        solutionInfo,
+                        razorDocument.Id,
+                        request.Position.ToLinePosition(),
+                        request.Character,
+                        autoCloseTags: enableAutoClosingTags,
+                        formatOnType: formatOnType,
+                        indentWithTabs: indentWithTabs,
+                        indentSize: indentSize,
+                        cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (data.Result is { } remoteInsertTextEdit)
+        {
+            _logger.LogDebug($"Got insert text edit from OOP {remoteInsertTextEdit}");
+            return RemoteInsertTextEdit.ToLspInsertTextEdit(remoteInsertTextEdit);
+        }
+
+        if (data.StopHandling)
+        {
+            return null;
+        }
+
+        // Got no data but no signal to stop handling, so try HTML
+        return await TryResolveHtmlInsertionAsync(razorDocument, request, cancellationToken)
+            .ConfigureAwait(false);
+    }
+
+    private async Task<VSInternalDocumentOnAutoInsertResponseItem?> TryResolveHtmlInsertionAsync(
+        TextDocument razorDocument,
+        VSInternalDocumentOnAutoInsertParams request,
+        CancellationToken cancellationToken)
+    {
+        // We support auto-insert in HTML only on "="
+        if (request.Character != "=")
```

Comment:
This should check `AutoInsertService.HtmlAllowedAutoInsertTriggerCharacters` rather than hard coding.

Also this method seems to be missing logic for the `AutoInsertAttributeQuotes` option

Reply:
Thanks, hardcoded equals was old code - we now check allowed characters in RemoteAutoInsertService and modify result based on that. So I just removed that check.

Missed that setting, added support for it now, thanks!